### ISSUE TITLE
Validation schema and online editor

### ISSUE THREAD

**Issue by simontaurus** [TheELNConsortium/TheELNFileFormat#67]
We are currently on the way implementing RO-Crate / ELN Fileformat for [OpenSemanticLab](https://github.com/OpenSemanticLab).

On the way we will create a validation schema using [OO-LD](https://github.com/OO-LD/schema).

As discussed with @SteffenBrinckmann this issue to share the work early and provide a first preview ([code](https://github.com/OO-LD/playground-rocrate)):
[playground example](https://oo-ld.github.io/playground-rocrate?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NigGYYAEwB00gOY2pYGxilkejYAjK7BfqQG4izozoSaLgpQghBU8cmpIAC%2BALpssFDYaSggAEoA8gC0AMIBekkpxexcPApSVACOIjDtcSYgifmNCjokMogAKmAKiM4ACiJ4uIiqUiB5IM5URDDwMLAIyGjgUh5kewBegXDwGiAZTWCc3NEg7V09VH2mD%2BtsED5UlJYDQ7r9juDFFApLsPAoqPARHRUKZyqdzjArgdbjksmx4GQGGDhiVENDYdk2CIpPIISSoTD4HC2GNZEYQNT5HiQBAyO1dGizpdrgg7gASdpEQgAYhsWx2e2xiBsngxWJu2W5kCVd2cMEQEAEgl4726vV4OEEqGhIioWW5OyKYP2TpKADEYE6hpkSk8Wq9EiVHWk2KbPt97sM2IkNgDIFRgTBQXSfaBIWTGcyQAikSiQB6nbi2LxA2n6RmKdyCUSU00K0zKZsaP4YNANbXCPW4dyEVg9Uy3dI2cSfQzK9G%2BmXR12FKzAoROQptTcjqARIawGRJxyN1veMpzs4imsSnoyQBVXfOAASh%2BPmrYuHgAGsjsZQMpJYRnEi6FbQ1QtIgAAbkmFAAAQ2OBzhgBQ8DYJucS5A6npUAATCOdbkg2%2BQuk0wYKAwepkHceC7HyggAKLwH2FIlHgZCIFQABsAAsCi8JCMB0F4EA4ZGo7AMAvZgP2HiDjIgT2o2FCBCodyLkGqGuIk3JPq%2BKKgO0QFYHQDC6MgbCftshBCZy0m5Gwy6HHcHgws4vBYNgSJIKgzFsHQIjYLAhrFDadr2lZbY2cceoGkaJqdGaXwWvBVooP5gWbIEjFUEYaZ4YQAAiKVMa8kJ%2Bi8bRReGea%2FBOfw8oCCYgquAnFFOWGZvCiLIv0OXkHl6zcqW9Wdth3b4oSDV9aSA2NlsSgwsFtwdmNzXcgeiBzHy6WjSAfIBP%2BxBDvOp5kHghkDHodB1QgVDlFKJigBKJklLK8q7PsK42AReI3V%2B91ytsT1KnKuVpdkuJJdZrmhfqhpCJFHzmpa1pSLa9rcnGQK1XcvXpuNbA5m1pg%2BJU%2FiBFQlQMOQzgpa4HC0Di3Lgo1%2FXNdjrV5rUVDXMBVAAOrSM%2B3VsKMQ6TNMc1jvxON5uYljWHYjguO4pzeL4hNBKEwR88leis0T26Y4zO2Sey5N6LADCVAhGByEuM11UQAiWJgz4nsASNsBzUiIO29PzeOijzIsyyrOKn0gA9P2Ki9qrCtijYHaIa23VdIffQqz2HP9nWA%2F81WJsmsR3OL%2FSuDY3XcvKZBeWtS0rcCKLA1q1t3LwYWQ8aYaw3F8OI9yKywbwCY%2BO7hC7HoAQ8Bq%2BSqDWIB4GAYBQBmZAQOxbBkBwZD6J3VBsM%2BVBUBAsVUGARC8MBci2kcRByExVJXgpV77neCaEGeUCXghW63vAR5P0lVOiscgY3hgAJnUYmpMtwUz%2Frcb0TQtawA5tzKQvN%2BYIFZELQgksrC2HsE4NwnhFZ%2BFASEMICgjZUDgV8QgCg3Ye3%2FiAGYfslj6kDiLQUaoRS3G5LHMQ6M%2BggCLg0UcHVUqvCrqtWuSMgA%3D%3D%3D)

**Comment by SteffenBrinckmann** [comment]
@simontaurus I like the idea: immediately show the content of the .eln file in a github.io and allow the user to inspect it.

**Comment by SteffenBrinckmann** [comment]
Hey,
I want to come back to the schema again, as it can help us clarify if something is complying. The ro-crate-metadata is a json and we should be able to create a schema for it. I have no experience in writing schema, but I am volunteering to create a github action to verify all .eln once the schema exists. 

@FlorianRhiem, @nicobrandt, @NicolasCARPi : can you help; have you already created one?  

There is no gui / preview aspect to this.

**Comment by NicolasCARPi** [comment]
TBH I'm surprised RO-Crate doesn't provide a schema. Seems it would be most helpful to everyone. This seems to discuss it: https://github.com/ResearchObject/ro-crate/issues/33

Maybe the JSON-LD nature of it makes it hard to create... Given that all the nodes can accept a wide range of properties, which can themselves have values of different types and subtypes, I wonder if it's even possible to create a schema that will validate all eln. It will necessarily be restrictive/incomplete, and we will need to adjust it often, based on our use.

Here is a generated one from an elabftw metadata, as a starting point:

~~~json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "@context": {
      "type": "string",
      "format": "uri"
    },
    "@graph": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "@id": {
            "type": "string"
          },
          "@type": {
            "type": "string"
          },
          "about": {
            "type": "object",
            "properties": {
              "@id": {
                "type": "string"
              }
            }
          },
          "conformsTo": {
            "type": "object",
            "properties": {
              "@id": {
                "type": "string",
                "format": "uri"
              }
            }
          },
          "dateCreated": {
            "type": "string",
            "format": "date-time"
          },
          "sdPublisher": {
            "type": "object",
            "properties": {
              "@id": {
                "type": "string"
              }
            }
          },
          "version": {
            "type": "string"
          },
          "author": {
            "type": "object",
            "properties": {
              "@id": {
                "type": "string"
              }
            }
          },
          "dateModified": {
            "type": "string",
            "format": "date-time"
          },
          "name": {
            "type": "string"
          },
          "encodingFormat": {
            "type": "string"
          },
          "url": {
            "type": "string",
            "format": "uri"
          },
          "genre": {
            "type": "string"
          },
          "creativeWorkStatus": {
            "type": "string"
          },
          "identifier": {
            "type": "string"
          },
          "keywords": {
            "type": "string"
          },
          "step": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "@type": {
                  "type": "string"
                },
                "position": {
                  "type": "integer"
                },
                "creativeWorkStatus": {
                  "type": "string"
                },
                "itemListElement": {
                  "type": "array",
                  "items": {
                    "type": "object",
                    "properties": {
                      "@type": {
                        "type": "string"
                      },
                      "text": {
                        "type": "string"
                      }
                    }
                  }
                }
              }
            }
          },
          "hasPart": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "@id": {
                  "type": "string"
                }
              }
            }
          },
          "comment": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "@id": {
                  "type": "string"
                },
                "@type": {
                  "type": "string"
                },
                "dateCreated": {
                  "type": "string",
                  "format": "date-time"
                },
                "text": {
                  "type": "string"
                },
                "author": {
                  "type": "object",
                  "properties": {
                    "@id": {
                      "type": "string"
                    }
                  }
                }
              }
            }
          }
        },
        "required": ["@id", "@type"]
      }
    }
  },
  "required": ["@context", "@graph"]
}
~~~

**Comment by simontaurus** [comment]
The main issue is that RO-CRATE specifies in principle an RDF graph, serialized as flattened and compacted JSON-LD (@graph with a list of nodes). This makes any syntactical validation (like JSON-SCHEMA) limited, in comparison to semantic RDF-"SCHEMAS" like SHACL (which, vise verse, have their limits in adaption and tool availability)

However, looking more closely, the RO-CRATE spec is not pure semantically but also syntactically: 
e.g. 
> @type: MUST be [Dataset](http://schema.org/Dataset)
> datePublished: MUST be a string in [ISO 8601 date format](http://schema.org/DateTime) and SHOULD be specified to at least the precision of a day, MAY be a timestamp down to the millisecond.

Pure semantically would mean, e.g. "the triple `<./.> <rdf:type> <http://schema.org/Dataset>` is present".
This leads to the interpretation (and implementations, also in https://github.com/TheELNConsortium/TheELNFileFormat/blob/master/tests/test_01_params_metadata_json.py#L53) that the named keys (e.g. `@type`) must be present but in fact this two JSON-LD files are semantically equivalent:
```jsonld
{
  "@context": "https://w3id.org/ro/crate/1.1/context",
  "@type": "Dataset"
}
```
```jsonld
{
  "@context": [
    "https://w3id.org/ro/crate/1.1/context",
    {
      "type": "@type"
    }
  ],
  "type": "Dataset"
}
```

In addition, RO-Crate inherits the very lax handling of data type, cardinality, range and required-ness of properties further comlicating validation and consumer implementation.
Therefore we would suggest to create at least for the ELN format syntactically (but sematically annotated) schemas (see [OO-LD](https://github.com/OO-LD/schema)) for a normalized tree-shape of the ro-create-metadata.json. Luckily JSON-LD can help use here with simply applying the `framing` algorithm, in combination with some aliases (`@id` => `id`, `@type` => `type`) in oder to get valid variable names e.g. in python.

<details>
<summary>Early ELN-Fileformat Schema Draft</summary>

```json
{
  "@context": [
    "https://w3id.org/ro/crate/1.1/context",
    {
      "id": "@id",
      "type": "@type"
    }
  ],
  "title": "RO-Crate",
  "type": "object",
  "version": "0.1.1",
  "required": [
    "id",
    "type",
    "conformsTo",
    "sdPublisher"
  ],
  "definitions": {
    "Thing": {
      "type": "object",
      "required": [
        "type"
      ],
      "properties": {
        "type": {
          "type": "string",
          "default": "Thing"
        },
        "id": {
          "type": "string"
        },
        "name": {
          "type": "string"
        },
        "description": {
          "type": "string"
        }
      }
    },
    "Organization": {
      "allOf": [
        {
          "$ref": "#/definitions/Thing"
        }
      ],
      "type": "object",
      "required": [
        "type"
      ],
      "properties": {
        "type": {
          "type": "string",
          "default": "Organization"
        },
        "url": {
          "type": "string",
          "format": "url"
        },
        "areaServed": {
          "type": "string"
        },
        "slogan": {
          "type": "string"
        },
        "logo": {
          "type": "string",
          "format": "url",
          "links": [
            {
              "href": "{{self}}",
              "type": "img/png"
            }
          ]
        },
        "parentOrganization": {
          "$ref": "#/definitions/Organization"
        }
      },
      "options": {
        "display_required_only": true
      }
    },
    "Person": {
      "allOf": [
        {
          "$ref": "#/definitions/Thing"
        }
      ],
      "title": "Person",
      "type": "object",
      "properties": {
        "email": {
          "type": "string",
          "format": "email"
        },
        "familyName": {
          "type": "string"
        },
        "givenName": {
          "type": "string"
        }
      }
    },
    "CreativeWork": {
      "allOf": [
        {
          "$ref": "#/definitions/Thing"
        }
      ],
      "title": "CreativeWork",
      "type": "object",
      "properties": {
    "dateCreated": {
      "type": "string",
      "format": "datetime-local",
      "options": {
        "flatpickr": {}
      }
    },
    "dateModified": {
      "type": "string",
      "format": "datetime-local",
      "options": {
        "flatpickr": {}
      }
    },
        "keywords": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    },
    "File": {
      "allOf": [
        {
          "$ref": "#/definitions/CreativeWork"
        }
      ],
      "title": "File",
      "type": "object",
      "id": "file",
      "required": [
        "type",
        "id"
      ],
      "properties": {
        "type": {
          "type": "string",
          "enum": [
            "File"
          ]
        },
        "_id": {
          "type": "string"
        },
        "encodingFormat": {
          "type": "string"
        },
        "id": {
          "type": "string",
          "format": "url",
          "options": {
            "upload": {
              "upload_handler": "testUploadHandler"
            }
          },
          "links": [
            {
              "href": "dummy",
              "rel": "view / download"
            }
          ]
        }
      },
      "not": { "required": [ "hasPart" ] },
      "options": {
        "display_required_only": true
      }
    },
    "Dataset": {
      "allOf": [
        {
          "$ref": "#/definitions/CreativeWork"
        }
      ],
      "title": "Dataset",
      "type": "object",
      "required": [
        "type",
        "id"
      ],
      "properties": {
        "type": {
          "type": "string",
          "enum": [
            "Dataset"
          ]
        },
        "about": {
          "$ref": "#/definitions/Thing"
        },
        "hasPart": {
          "type": "array",
          "format": "tabs",
          "items": {
        "discriminator": {
        "propertyName": "type",
        "mapping": {
          "File": "#/definitions/File",
          "Dataset": "#/definitions/Dataset"
        }
      },
            "oneOf": [
              {
                "$ref": "#/definitions/File"
              },
              {
                "$ref": "#/definitions/Dataset"
              }
            ]
          }
        }
      },
      "options": {
        "display_required_only": true
      }
    }
  },
  "properties": {
    "id": {
      "type": "string",
      "enum": [
        "ro-crate-metadata.json"
      ]
    },
    "type": {
      "type": "string",
      "enum": [
        "CreativeWork"
      ]
    },
    "conformsTo": {
      "type": "string",
      "enum": [
        "https://w3id.org/ro/crate/1.1"
      ]
    },
    "version": {
      "type": "string"
    },
    "sdPublisher": {
      "$ref": "#/definitions/Organization"
    },
    "about": {
      "$ref": "#/definitions/Dataset",
      "properties": {
        "id": {
          "enum": [
            "./"
          ]
        }
      },
      "default": {
        "hasPart": []
      }
    }
  },
  "options": {
    "_display_required_only": true
  }
}
```

</details>
 
 This would provide use three outcomes:
 1. programming language agnostic validation by using JSON-LD normalization + JSON-SCHEMA
 1. code generation to automate the parsing and serialization
 1. UI generation for generic viewers / editors

As a demo for 1. and 3., 
[ELN Fileformat Playground - PASTA.eln ](https://oo-ld.github.io/playground-rocrate/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NigGYYAEwB00gOY2pYGxilkejYAjK7BfqQG4izozoSaLgpQghBU8cmpIAC%2BALpssFDYaSggAEoA8gC0AMIBekkpxexcPAoAblRSiHDwhAAMYWEKUlQAjiIwI3EmIIn5jQo6JDKIACpgCojOAAoieLiIqlIgeSDOVEQw8DCwCMhoIKvKVx4aIBlNYJzc0SAj45MqNNTB8TmwID5UlJYDQ3qCHvDFFApC8FOciGQRNgjI9nvBXlk2IkEQsSohkajCSB4GQGHDSUiUfjsmxzkoUdAevTMmSKcysgK2OUpB4yNcAF6BLkPOTYcpEVDGUAAEhGCpKAGIbOirjceogbE9KadEV8Wr9%2FhMpor3gtThDIJ0YfdQPDXQzyUzXqyLpjsYRhaKJVKECyQCIpPISTzGai2MtZDiI%2FIqWQRmQAMqdDrTd0xz2UtiIbBgIPcpoF%2FlsEseDbRit873EaSJwjJhS4eAAa3uSrMasIwGAiCo2CIAoaMZgdC8EH5OSpEDTVF0gbFMElt16D1VF0IWp11y3BrXwa32SpkGPb2cMEQEAEgl4lsBzl4OEEqGRIiogpA206Wht1AWV5UVFUB01bULl1Y9DTxAkTRuIpCAArpQ3mGMzR%2BBQHShZ03jUMgYCjPMGy9BQE0CQgiJIsMMToEjBAAOVpYoyMISsCTYDwYA6eBWLpetOMbC8qVqKgpQ6AB1aQuzeUD1T7Xd1RAA8YKPfV4ONfJkKaCSpKoWSpHkzDPm%2BVpwUhJ0YFhB5nECKgDL0XNbXzUT4xbaiSgcvRYAYSoSwwOQFCvfU3iIARLEwLtjjQAUqV8qgAFkwFvS4gXLESKM8mRvLORz%2FKoQKwGC%2BQ2DCu4IqiiAYri4AErYLsqEEChpGcF03KaNMAk%2FIk9DoTrES4sSqQAMRI9iQEU8CQBU%2FdoMuTS7hsZy%2BKMuTsiQwomgmlCzMIbDLNmaZiEm4YxitTKZlBIk4ntazoVsobSQ43kcpAFcRDoG09rSBc2F4Yk3tjKtPvgLBb3xMavKMEGRqpYGuuyuNmzypNI1CzkqoeEQHzAMhXLxktCd4ZQxWcIo4veGgoAAVXxwmAAkKapi9qyuHtZuUSCzm%2Bug%2Br%2BUdCDaWyKAAAhscXnDACh4BJuJcj%2FSQ4aFgFrRmcnEG2NNxAB8BsaQG87wfIRn0u1933lz8UG%2FX8qQAEUCMgR1VmaTAgvcoMPPUVrWmTNtyXSdsIJ3yFdydzPNC71eukEFjusEQDwmy7Le%2BGPPB76bTDl2qD11M8FEVX5u9jTfaQbSwa1nXoSykoeqESjYcIcg8GQfq1E6292WnK5AmkN4U%2BhFi2NbhOQFkCBauZB4%2FoWn24Pnthc4jsulorg1V%2Fz9nwHgKgwI9ubefUjel%2FOwlPdU0%2FYK07fxCVv9KqN%2ByTcfc3Y7fD8vykH9GuTx6BEHhI2GpnL6P0Zg%2BEqP4RylQGDkF8mQVwHAgInCpG6ZG71UbgJtP7DaJk0FsCWC2NYdYM4fRwZrCwVhbD2CcG4Tw3hfAwKCKEYIhCQAdC6NKchhZFA7D2AcI4bxS5qUWrfFap4Nwhl6IXYuIiT7iOWpXe%2BuFAHPTeEjShphXA2DQYlX0WJVY111oqBcT9DadV4D3U2T4XxTCttgG2ds%2FyHFlrwToPguiECuHoAIPAuT5FUEJEARcwBQE9GQCAAAWBQZAOBkH0D%2FH8TUqBUAgFbKgYAiC8DaHIH89wMTYBHGwYmBMiaMzfOTeAlNOit1pgzBWLMalsz%2FCg0MwDTpQJYcVeBhNnbINQQdEoeDjKmRAMQlY6xCDmEsNYOwjgXDuBFEwvwdQqAhCGGwLh3QOkgE2AI%2FYd5hHCRKFIzcXI5FiE0adXRkdQ7OwjmwExdcTDmPjJNOmmMSizJoXYAIFBXC8SgMoPYIgRxSB0HoXQrgsB0HglQAAogAGWYtUO40hYDfQRSi5if0YboxsLIcknQbAGFpA%2BGgNhtgAEEMyrBpdSulDLXCjlkUAA&fileUrl=https://raw.githubusercontent.com/TheELNConsortium/TheELNFileFormat/master/examples/PASTA/PASTA.eln)
reveals that Steffen has nested a `Dataset` in a `File`, which could at least be discussed.

![grafik](https://github.com/user-attachments/assets/99cc8200-ff45-4d48-92b0-cd0ae7eefe7b)


For 2., throwing the same schema at [OO-LD Python playground](https://oo-ld.github.io/playground-python-yaml/?data=N4Ig9gDgLglmB2BnEAuUMDGCA2MBGqIAZglAIYDuApomALZUCsIANOHgFZUZQD62ZAJ5gArlELwwAJzplsrEIgwALKrNSgAAlnhQqAD3EoA2iGVQoERCgD0NigGYYAEwB00gOY2pYGxilkejYAjK7BfqQG4izozoSaLgpQghBU8cmpIAC%2BALpssFDYaSggAEoA8gC0AMIBekkpxexcPAoAblRSiHDwhAAMYWEKUlQAjiIwI3EmIIn5jQo6JDKIACpgCojOAAoieLiIqlIgeSDOVEQw8DCwCMhoIKvKVx4aIBlNYJzc0SAj45MqNNTB8TmwID5UlJYDQ3qCHvDFFApC8FOciGQRNgjI9nvBXlk2IkEQsSohkajCSB4GQGHDSUiUfjsmxzkoUdAevTMmSKcysgK2OUpB4yNcAF6BLkPOTYcpEVDGUAAEhGCpKAGIbOirjceogbE9KadEV8Wr9%2FhMpor3gtThDIJ0YfdQPDXQzyUzXqyLpjsYRhaKJVKECyQCIpPISTzGai2MtZDiI%2FIqWQRmQAMqdDrTd0xz2UtiIbBgIPcpoF%2FlsEseDbRit873EaSJwjJhS4eAAa3uSrMasIwGAiCo2CIAoaMZgdC8EH5OSpEDTVF0gbFMElt16D1VF0IWp11y3BrXwa32SpkGPb2cMEQEAEgl4lsBzl4OEEqGRIiogpA206Wht1AWV5UVFUB01bULl1Y9DTxAkTRuIpCAArpQ3mGMzR%2BBQHShZ03jUMgYCjPMGy9BQE0CQgiJIsMMToEjBAAOVpYoyMISsCTYDwYA6eBWLpetOMbC8qVqKgpQ6AB1aQuzeUD1T7Xd1RAA8YKPfV4ONfJkKaCSpKoWSpHkzDPm%2BVpwUhJ0YFhB5nECKgDL0XNbXzUT4xbaiSgcvRYAYSoSwwOQFCvfU3iIARLEwLtjjQAUqV8qgAFkwFvS4gXLESKM8mRvLORz%2FKoQKwGC%2BQ2DCu4IqiiAYri4AErYLsqEEChpGcF03KaNMAk%2FIk9DoTrES4sSqQAMRI9iQEU8CQBU%2FdoMuTS7hsZy%2BKMuTsiQwomgmlCzMIbDLNmaZiEm4YxitTKZlBIk4ntazoVsobSQ43kcpAFcRDoG09rSBc2F4Yk3tjKtPvgLBb3xMavKMEGRqpYGuuyuNmzypNI1CzkqoeEQHzAMhXLxktCd4ZQxWcIo4veGgoAAVXxwmAAkKapi9qyuHtZuUSCzm%2Bug%2Br%2BUdCDaWyKAAAhscXnDACh4BJuJcj%2FSQ4aFgFrRmcnEG2NNxAB8BsaQG87wfIRn0u1933lz8UG%2FX8qQAEUCMgR1VmaTAgvcoMPPUVrWmTNtyXSdsIJ3yFdydzPNC71eukEFjusEQDwmy7Le%2BGPPB76bTDl2qD11M8FEVX5u9jTfaQbSwa1nXoSykoeqESjYcIcg8GQfq1E6292WnK5AmkN4U%2BhFi2NbhOQFkCBauZB4%2FoWn24Pnthc4jsulorg1V%2Fz9nwHgKgwI9ubefUjel%2FOwlPdU0%2FYK07fxCVv9KqN%2ByTcfc3Y7fD8vykH9GuTx6BEHhI2GpnL6P0Zg%2BEqP4RylQGDkF8mQVwHAgInCpG6ZG71UbgJtP7DaJk0FsCWC2NYdYM4fRwZrCwVhbD2CcG4Tw3hfAwKCKEYIhCQAdC6NKchhZFA7D2AcI4bxS5qUWrfFap4Nwhl6IXYuIiT7iOWpXe%2BuFAHPTeEjShphXA2DQYlX0WJVY111oqBcT9DadV4D3U2T4XxTCttgG2ds%2FyHFlrwToPguiECuHoAIPAuT5FUEJEARcwBQE9GQCAAAWBQZAOBkH0D%2FH8TUqBUAgFbKgYAiC8DaHIH89wMTYBHFSFBoZgGnSgSw4q8DCbO2Qagg6JQ8HGVMiAYhKx1iEHMJYawdhHAuHcCKJhfg6hUBCEMNgXDujlJAJsAR%2Bw7zCOEiUKRm4uRyLEJo06ujI6h2dhHNgJi64mHMVkIAA%3D) gives us pydantic dataclasses both for validation and implementation.

<details>
<summary>Generated Dataclasses</summary>

```python
from __future__ import annotations

from enum import Enum
from typing import List, Literal, Optional, Union

from pydantic import BaseModel, EmailStr, Field


class Id(Enum):
    ro_crate_metadata_json = "ro-crate-metadata.json"


class Type(Enum):
    CreativeWork = "CreativeWork"


class ConformsTo(Enum):
    https___w3id_org_ro_crate_1_1 = "https://w3id.org/ro/crate/1.1"


class Thing(BaseModel):
    type: str
    id: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None


class Organization(Thing):
    type: str
    url: Optional[str] = None
    areaServed: Optional[str] = None
    slogan: Optional[str] = None
    logo: Optional[str] = Field(None, links=[{"href": "{{self}}", "type": "img/png"}])
    parentOrganization: Optional[Organization] = None


class Person(Thing):
    email: Optional[EmailStr] = None
    familyName: Optional[str] = None
    givenName: Optional[str] = None


class CreativeWork(Thing):
    dateCreated: Optional[str] = Field(None, options={"flatpickr": {}})
    dateModified: Optional[str] = Field(None, options={"flatpickr": {}})
    keywords: Optional[List[str]] = None


class Type1(Enum):
    File = "File"


class File(CreativeWork):
    type: Literal["File"]
    field_id: Optional[str] = Field(None, alias="_id")
    encodingFormat: Optional[str] = None
    id: str = Field(
        ...,
        links=[{"href": "dummy", "rel": "view / download"}],
        options={"upload": {"upload_handler": "testUploadHandler"}},
    )


class Type2(Enum):
    Dataset = "Dataset"


class ROCrate(BaseModel):
    id: Id
    type: Type
    conformsTo: ConformsTo
    version: Optional[str] = None
    sdPublisher: Organization
    about: Optional[Dataset] = Field(
        default_factory=lambda: Dataset.parse_obj({"hasPart": []})
    )


class Dataset(CreativeWork):
    type: Literal["Dataset"]
    about: Optional[Thing] = None
    hasPart: Optional[List[Union[File, Dataset]]] = Field(None, discriminator="type")
    id: str
```

</details>

While this approach never forbidds additional properties we can easily define in a machine readable way which properties we expect to be used and how we expect them to be used. Also we can define subclasses of `Dataset` or `File` that have a strict definition of metadata in order to allow automated processing, e.g. `CsvFile` with [CSVW](https://w3c.github.io/csvw/syntax/) conform column annotations.

What do you think?

**Comment by SteffenBrinckmann** [comment]
@simontaurus How do we continue here, this issue seems  stalled.

**Comment by simontaurus** [comment]
#116 and #98 closed makes it easier to continue here  - I will work on it

**Comment by ptsefton** [comment]
How did the framing efforts go? I tried this in the early development on RO-Crate and I could never get it to work relliably -- even setting aside issues with cycles in the graph.

**Comment by SteffenBrinckmann** [comment]
Is this the approach that allotrope takes? They should have the same challenge.

**Comment by NicolasCARPi** [comment]
@simontaurus think we can close this issue? Validator: https://check-eln.streamlit.app/

**Comment by CAOShurong** [comment]
I tested the checker against current reference archives before deciding whether the validator is enough to close this issue.

Scope: the four check functions in [`checkELN` at `2bdff1b`](https://github.com/SteffenBrinckmann/checkELN/tree/2bdff1baa9047baa0b8fc590aa7db098f8fb08c8), compared with the same four functions in [TheELNFileFormat at `19028dd`](https://github.com/TheELNConsortium/TheELNFileFormat/tree/19028dd2878065b34df7a7df836babaa61ad2845). I ran them headlessly with the checker's pinned dependencies against all 11 `.eln` archives currently under `examples/`; I did not automate the Streamlit UI.

Two reproducible gaps look relevant:

1. **The checker copy has drifted from the current conformance checks.** For the current [SciLog example](https://github.com/TheELNConsortium/TheELNFileFormat/blob/19028dd2878065b34df7a7df836babaa61ad2845/examples/SciLog/export%20-%202026-06-05%2003_25_10%20GMT%2B2.eln) (SHA-256 `2a264b20683f3c30aeb10ef7e4223fe8d7a141505e41855ddfc52244e90f43a0`), `checkELN` returned `pass / fail / pass / fail` for PyPI RO-Crate / validator / params / schema. The current TheELN checks returned `pass / pass / pass / pass` for the same file in the same environment. The schema failures are the array forms of `@context` and `@type`: the [checker schema accepts only strings](https://github.com/SteffenBrinckmann/checkELN/blob/2bdff1baa9047baa0b8fc590aa7db098f8fb08c8/schema.json#L5-L18), while the [current consortium schema accepts strings or arrays](https://github.com/TheELNConsortium/TheELNFileFormat/blob/19028dd2878065b34df7a7df836babaa61ad2845/tests/schema.json#L5-L18). The validator failure was `ro-crate-1.1_5.3`; the [current check has the RO-Crate 1.2 compatibility handling for SciLog and SampleDB](https://github.com/TheELNConsortium/TheELNFileFormat/blob/19028dd2878065b34df7a7df836babaa61ad2845/tests/checks.py#L194-L205), which the checker copy lacks.

2. **Neither implementation checks the mandatory outer archive layout.** Starting from the current OpenSemanticLab minimal example (SHA-256 `efe2ac320d754f81d75691c7325d192b3f0300043200ffe09491c36760ae739c`), I appended one root-level entry named `unexpected-root.txt`. The resulting archive has two top-level entries, `MinimalExample` and `unexpected-root.txt`, so it violates the specification's [single-root-folder MUST](https://github.com/TheELNConsortium/TheELNFileFormat/blob/19028dd2878065b34df7a7df836babaa61ad2845/SPECIFICATION.md#L12-L16). All four checks returned success in both implementations. The derived archive SHA-256 is `e9f329deb4e592c1b2ed9e6b7215d953468423eaf4264f28280184c749efbcc8`; it can be reproduced with:

```python
from pathlib import Path
from shutil import copyfile
from zipfile import ZipFile

src = Path("examples/OpenSemanticLab/MinimalExample.osl.eln")
dst = Path("opensemanticlab-extra-root-entry.eln")
copyfile(src, dst)
with ZipFile(dst, "a") as archive:
    archive.writestr("unexpected-root.txt", "x")
```

The online checker is useful, but these false-negative and false-positive cases mean I would not treat it as a complete `.eln` conformance gate yet. Keeping this issue open until the app consumes the current shared checks/schema and validates the outer ZIP layout would give it a concrete, testable completion criterion.

**Comment by CAOShurong** [comment]
Follow-up after #156 merged into `master` as [`73286db7`](https://github.com/TheELNConsortium/TheELNFileFormat/commit/73286db7de6e3802db1354a1487a67b65b27e2d1): I audited whether the remaining checker drift can safely be fixed directly in [`SteffenBrinckmann/checkELN`](https://github.com/SteffenBrinckmann/checkELN/tree/2bdff1baa9047baa0b8fc590aa7db098f8fb08c8). I did **not** open a PR there.

The blocking evidence is concrete: GitHub reports no detected license and its repository-license endpoint returns 404; the repository has no `LICENSE`, `CONTRIBUTING`, or README, no CI workflow, no issue or PR history, only one branch, and no push since 2025-09-24. Public/unarchived status alone does not clarify redistribution rights or whether outside patches are accepted. Also, unauthenticated requests to both the deployed app root and `/_stcore/health` currently return HTTP 303 redirects to Streamlit authentication, so I could not verify the public UI path headlessly.

The smallest implementation once those repository-level boundaries are clear would be:

1. vendor the MIT-licensed [`tests/checks.py`](https://github.com/TheELNConsortium/TheELNFileFormat/blob/73286db7de6e3802db1354a1487a67b65b27e2d1/tests/checks.py) and [`tests/schema.json`](https://github.com/TheELNConsortium/TheELNFileFormat/blob/73286db7de6e3802db1354a1487a67b65b27e2d1/tests/schema.json) at that exact commit, with source, commit, MIT notice, and hashes;
2. update `rocrate` from 0.13.0 to the consortium's 0.14.0 and explicitly declare the Streamlit runtime dependency;
3. write the Streamlit upload once to a temporary filesystem path before calling the shared functions—the current app passes an uploaded-file object, but the shared `checkValidator` now calls `Path(fileName)`, so blind file replacement would break that path;
4. expose all five shared checks, including `checkArchiveStructure`;
5. add a pinned-provenance drift test plus license-safe behavioral tests for array-valued `@context`/`@type` and a generated extra-root ZIP.

I checked the narrow design without retaining any third-party archive: a generated array-valued JSON-LD crate fails checkELN's old schema and passes the `73286db7` schema; the shared archive check accepts its single-root form and rejects a generated variant containing `unexpected-root.txt`. I did not rerun or retain the SciLog archive in this audit because its own metadata states `No license`; the earlier comment above remains the evidence for that historical reproduction, not a new run here.

That leaves the technical path small, but the correct stopping point is still the repository license/contribution boundary rather than an unsolicited code PR.

**Comment by NicolasCARPi** [comment]
@SteffenBrinckmann we should merge checkeln in this repo!

**Comment by CAOShurong** [comment]
Yes - I can take a bounded first implementation in this repository.

To avoid copying code from the currently unlicensed `SteffenBrinckmann/checkELN` repository, I would make this a clean implementation against this repository's MIT-licensed `tests/checks.py` and `tests/schema.json` as the single source of truth:

- a small Streamlit upload entry point in this repository;
- one temporary filesystem path between the uploaded file and the five shared checks, including `checkArchiveStructure`;
- regression coverage for array-valued `@context` / `@type`, a generated extra-root archive, and the upload-to-path bridge;
- declared runtime dependencies and deployment instructions.

If that scope and placement match what you have in mind, I will open one focused PR. @SteffenBrinckmann, if any existing `checkELN` code or assets should be preserved instead of cleanly reimplemented, please confirm what may be reused and under which license.

Prepared with OpenAI Codex assistance; I reviewed the repository contribution/license boundaries and the implementation plan above.

**Comment by SteffenBrinckmann** [comment]
Copy paste away, the tester in MIT license, too 

I want to update the entire test suite sometime in the next months given that ro crate increases the spec version faster than ever before.

**Comment by SteffenBrinckmann** [comment]
The schema-validation work and the archive-structure check are now implemented in the repository. The remaining checker-integration work is tracked in #157.

The OO-LD framing and generated-editor exploration is not currently planned (It might be picked up in the future by @simontaurus)

I suggest closing this issue as it combines too many different topics. @NicolasCARPi
